File: platform/app/quote_service.py
"""Quote service — session-scoped create / get / update (brief §6, §9, §17.4).

The backend owns the journey: it stores quote state keyed to a strong-entropy
session id, deep-merges partial greedy patches (never blanking siblings),
recomputes ``missingFields`` and ``journeyState`` server-side, and emits domain
events so the dashboard and audit trail come for free (three-layer discipline).

State shape returned to callers (brief §6):
    { quoteId, journeyState, missingFields, currentOutcome }
``sessionId`` is returned **only at creation** — never echoed by get/patch.
"""

from __future__ import annotations

from typing import Any, Optional

from app.events import store as event_store
from app.required import missing_fields
from app.sessions import QuoteRecord
from app.sessions import store as session_store


def _journey_state(data: dict, missing: list[str]) -> str:
    """Derive the journey state from the stored data (brief §6).

    - ``quote_started``  — nothing collected yet.
    - ``collecting``     — some data, still missing mandatory fields.
    - ``ready_to_price`` — no mandatory fields remain.
    """
    if not missing:
        return "ready_to_price"
    if not data:
        return "quote_started"
    return "collecting"


def _state(record: QuoteRecord) -> dict:
    """Build the brief §6 state object. Never includes the sessionId."""
    missing = missing_fields(record.data)
    return {
        "quoteId": record.quote_id,
        "journeyState": _journey_state(record.data, missing),
        "missingFields": missing,
        "currentOutcome": None,
    }


def _is_empty_leaf(value: Any) -> bool:
    """Null/empty leaves are dropped before merging (brief §17.4)."""
    if value is None:
        return True
    if isinstance(value, str) and value.strip() == "":
        return True
    return False


def deep_merge(base: dict, patch: dict) -> dict:
    """Deep-merge ``patch`` into ``base`` in place (brief §17.4).

    - Nested dicts merge recursively, so a greedy patch touching one leaf never
      blanks its siblings.
    - Null / empty-string leaves in the patch are dropped (never used to blank
      existing data).
    - Lists (e.g. ``namedDrivers``) replace wholesale — there is no leaf to
      preserve, and named-driver append has its own path in a later slice.
    """
    for key, value in patch.items():
        if isinstance(value, dict):
            existing = base.get(key)
            if not isinstance(existing, dict):
                existing = {}
            merged = deep_merge(existing, value)
            # Only set if the merge produced something (avoid blank sub-objects).
            if merged:
                base[key] = merged
        elif _is_empty_leaf(value):
            # Drop — never blank an existing value with null/empty.
            continue
        else:
            base[key] = value
    return base


def create_quote() -> dict:
    """Create a draft quote bound to a fresh session.

    Emits ``QUOTE_CREATED`` (domain). The payload includes the quoteId but
    **never** the sessionId.
    """
    record = session_store.create()
    event_store.append("QUOTE_CREATED", {"quoteId": record.quote_id}, "domain")
    state = _state(record)
    # sessionId is returned only here, at creation.
    return {**state, "sessionId": record.session_id}


def get_quote(quote_id: str, session_id: str) -> Optional[dict]:
    """Return the state for ``quote_id`` iff ``session_id`` matches; else None."""
    record = session_store.get(quote_id, session_id)
    if record is None:
        return None
    return _state(record)


def apply_patch(quote_id: str, session_id: str, patch: dict) -> Optional[dict]:
    """Deep-merge ``patch`` into the quote; recompute state; emit QUOTE_UPDATED.

    Returns ``None`` on unknown quote or session mismatch (treated as not-found).
    """
    record = session_store.get(quote_id, session_id)
    if record is None:
        return None
    deep_merge(record.data, patch or {})
    event_store.append("QUOTE_UPDATED", {"quoteId": record.quote_id}, "domain")
    return _state(record)

File: platform/src/main/java/com/acme/platform/quote/QuoteService.java
package com.acme.platform.quote;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.springframework.stereotype.Service;

import com.acme.platform.events.EventStore;

/**
 * Quote service — session-scoped create / get / update (brief §6, §9, §17.4).
 *
 * <p>The backend owns the journey: it stores quote state keyed to a
 * strong-entropy session id, deep-merges partial greedy patches (never blanking
 * siblings), recomputes {@code missingFields} and {@code journeyState}
 * server-side, and emits domain events so the dashboard and audit trail come
 * for free (three-layer discipline).
 *
 * <p>State shape returned to callers (brief §6):
 * {@code {quoteId, journeyState, missingFields, currentOutcome}}.
 * {@code sessionId} is returned <b>only at creation</b> — never by get/patch.
 */
@Service
public class QuoteService {

    private final SessionStore sessions;
    private final EventStore events;

    public QuoteService(SessionStore sessions, EventStore events) {
        this.sessions = sessions;
        this.events = events;
    }

    /** Create a draft quote bound to a fresh session. Emits {@code QUOTE_CREATED}. */
    public Map<String, Object> createQuote() {
        QuoteRecord record = sessions.create();
        events.append("QUOTE_CREATED", Map.of("quoteId", record.quoteId()), "domain");
        Map<String, Object> state = state(record);
        // sessionId is returned only here, at creation.
        state.put("sessionId", record.sessionId());
        return state;
    }

    /** Return the state for {@code quoteId} iff {@code sessionId} matches; else {@code null}. */
    public Map<String, Object> getQuote(String quoteId, String sessionId) {
        QuoteRecord record = sessions.get(quoteId, sessionId);
        if (record == null) {
            return null;
        }
        return state(record);
    }

    /**
     * Deep-merge {@code patch} into the quote; recompute state; emit
     * {@code QUOTE_UPDATED}. Returns {@code null} on unknown quote or session
     * mismatch (treated as not-found).
     */
    public Map<String, Object> applyPatch(String quoteId, String sessionId, Map<String, Object> patch) {
        QuoteRecord record = sessions.get(quoteId, sessionId);
        if (record == null) {
            return null;
        }
        deepMerge(record.data(), patch == null ? Map.of() : patch);
        events.append("QUOTE_UPDATED", Map.of("quoteId", record.quoteId()), "domain");
        return state(record);
    }

    /** Build the brief §6 state object. Never includes the sessionId. */
    private Map<String, Object> state(QuoteRecord record) {
        List<String> missing = RequiredFields.missingFields(record.data());
        Map<String, Object> state = new LinkedHashMap<>();
        state.put("quoteId", record.quoteId());
        state.put("journeyState", journeyState(record.data(), missing));
        state.put("missingFields", missing);
        state.put("currentOutcome", null);
        return state;
    }

    /**
     * Derive the journey state (brief §6):
     * {@code quote_started} (nothing collected) →
     * {@code collecting} (some data, gaps) →
     * {@code ready_to_price} (no mandatory fields remain).
     */
    private static String journeyState(Map<String, Object> data, List<String> missing) {
        if (missing.isEmpty()) {
            return "ready_to_price";
        }
        if (data == null || data.isEmpty()) {
            return "quote_started";
        }
        return "collecting";
    }

    /**
     * Deep-merge {@code patch} into {@code base} in place (brief §17.4).
     *
     * <ul>
     *   <li>Nested maps merge recursively, so a greedy patch touching one leaf
     *       never blanks its siblings.</li>
     *   <li>Null / empty-string leaves in the patch are dropped (never used to
     *       blank existing data).</li>
     *   <li>Lists (e.g. {@code namedDrivers}) replace wholesale.</li>
     * </ul>
     */
    @SuppressWarnings("unchecked")
    static Map<String, Object> deepMerge(Map<String, Object> base, Map<String, Object> patch) {
        for (Map.Entry<String, Object> entry : patch.entrySet()) {
            String key = entry.getKey();
            Object value = entry.getValue();
            if (value instanceof Map<?, ?> patchMap) {
                Object existing = base.get(key);
                Map<String, Object> target = (existing instanceof Map)
                    ? (Map<String, Object>) existing
                    : new LinkedHashMap<>();
                Map<String, Object> merged = deepMerge(target, (Map<String, Object>) patchMap);
                // Only set if the merge produced something (avoid blank sub-objects).
                if (!merged.isEmpty()) {
                    base.put(key, merged);
                }
            } else if (isEmptyLeaf(value)) {
                // Drop — never blank an existing value with null/empty.
            } else {
                base.put(key, value);
            }
        }
        return base;
    }

    /** Null/empty leaves are dropped before merging (brief §17.4). */
    private static boolean isEmptyLeaf(Object value) {
        if (value == null) {
            return true;
        }
        return value instanceof String s && s.strip().isEmpty();
    }
}
